### ISSUE TITLE
cumsum() could (should?) return a summed area table for higher dimensional inputs

### ISSUE THREAD

**Issue by anntzer** [numpy/numpy#7993]
Currently, cumsum() operates on flattened arrays.  This behavior is not that useful, as even in its absence it could trivially be obtained by calling ravel() yourself first.
Perhaps a more useful approach would be to return a summed area table (https://en.wikipedia.org/wiki/Summed_area_table), i.e. compute for each position the sum of entries with no indices greater than the current entry, which is AFAICT trickier to compute in a "vectorized" manner.
I guess a reasonable transition path would be to start by emitting a warning whenever cumsum() is called on a >=2d array with no axis kwarg.

**Comment by jaimefrio** [comment]
Such a behavior change would require a very long and cumbersome deprecation period, and you are going to have a hard time convincing people that the benefits are worth the pain. I for one don't think they are...

What would be simpler and almost as nice is if `cumsum` would take a tuple as an `axis` parameter. Not sure if this change should be pursued in the `accumulate` method of ufuncs (there probably has to be a commutativity requirement on the operator for the multidimensional accumulation to make sense, right?) or be implemented in `cumsum` (and maybe `cumprod`?) only.

A related issue that has been discussed before, and which is especially relevant for getting usable summed area tables, is adding some way of telling `accumulate` that you want the operator's null element as the first item of the return array: if you have to manually add a leading row and column of zeros to the output of this enhanced `cumsum`, you might as well do the multiple calls to `cumsum` yourself...

**Comment by anntzer** [comment]
Yes, this only makes sense for commutative operators.  In fact, you probably also need elements to have an inverse for the operator for the table to be really useful.  Altogether I'd say implementing this for `accumulate` is a bit overkill.

Even regardless of deprecation policy, I think your remark about adding the null element in from of the array is the real argument here in favor of a new function (alternatively, adding a new kwarg to `cumsum`, but I prefer the former).

**Comment by shoyer** [comment]
The default flattening behavior of many NumPy operations a perpetual annoyance. I do agree with @jaimefrio though.

See #6044 for a discussion of adding the identity element.

**Comment by anntzer** [comment]
Somewhat ironically, I just hit a case where I needed flattening behavior on `bincount`, which does *not* provide such semantics.  Although I easily worked around this by raveling the array, I noticed that at least, even if many functions default to flattening, they should at least be able to do so without having to ravel the data in memory (which will imply a copy if the the initial nd array was e.g. a slice of a larger array).

Unfortunately it seems that cumsum doesn't even provide this optimization -- the in-memory raveling seems to be done by `PyArray_CheckAxis`, which calls `PyArray_Ravel`.
